### ISSUE TITLE
Map refresh not working properly.

### ISSUE THREAD

**Issue by bashbers** [pauldendulk/Mapsui#251]
We've created a custom user control with dependency properties to bind on the layers. The reason we did this is because to comply with the MVVM pattern. 

The first time we load the page, the map is drawn correctly. However whenever we navigate away and return to the page, the map isn't drawn properly. 

![screenshot_4](https://user-images.githubusercontent.com/23037527/32886858-60f57748-cac2-11e7-92c9-d5043ed596ee.png)
![screenshot_1](https://user-images.githubusercontent.com/23037527/32886855-60b0b4c8-cac2-11e7-81f8-c81fa41e88c2.png)
![screenshot_2](https://user-images.githubusercontent.com/23037527/32886856-60c6ddca-cac2-11e7-820e-b7c333995459.png)
![screenshot_3](https://user-images.githubusercontent.com/23037527/32886857-60db62a4-cac2-11e7-97dc-16c9937c3fa5.png)

We've added some pictures for reference. Is it possible to create a dependencyproperty in the usercontrol itself or do we need to keep this custom made usercontrol? The view/pagecontrol is unloaded. However the list of layers keeps existing. We've tried to dispose it and reinitialize it. This does not have any effect on how the map is drawn from the second time forward. We do call refresh() the moment the datasource has changed or the layerlist have been changed.

Is this issue known and if so, how do we resolve it?

**Comment by pauldendulk** [comment]
This is not a known issue. I would expect an update after Map.Refresh.

What happens if you drag in the Map?

Could it be you somehow create a copy of the MapControl when navigating? Or a copy of the Map object and you are trying to refresh the old Map object.

**Comment by pauldendulk** [comment]
And thanks for the screenshots, it definitely helps

**Comment by bashbers** [comment]
Whenever I drag the map, the map flickers between the full map and the partial map as posted. 

The only thing that persists between switching the tabs is the layerslist. 
The whole page is recreated and on that page is a MapControl.

**Comment by pauldendulk** [comment]
The 'partial map', is that the map view of Amsterdam, of the view where part of the map is visible (last screenshot).

**Comment by pauldendulk** [comment]
Perhaps the old map is still dangling somewhere and messing things up. What if you first remove the layers from the old map before you switch away?

**Comment by bashbers** [comment]
We execute this method every time the collection of layers has changed and when the property is initialized.

```c#
        private void RefreshMap()
        {
            this.MainMapControl.Map.Layers.Clear();
            foreach (var baseLayer in this.Layers)
            {
                this.MainMapControl.Map.Layers.Add(baseLayer);
            }
            this.MainMapControl.Refresh();
}
```

**Comment by bashbers** [comment]
Thanks for responding so fast by the way!

**Comment by pauldendulk** [comment]
Well, I don't know exactly what is going but I think you should try to recreate either everything from scratch when navigating back or just keep everything in memory (although it should be enough to preserve just the Map object).

**Comment by pauldendulk** [comment]
You could mail the the project at pauldendulk@gmail.com. I am curious about what is going on.

**Comment by bashbers** [comment]
Thanks to @pauldendulk we found a workaround. If someone else is experiencing a similar issue, change the MapControl.rendermode to rendermode.skia.

```c#

       public MapControl()
        {
            InitializeComponent();
            MainMapControl.Map.CRS = "EPSG:28992";
            MainMapControl.RenderMode = Mapsui.UI.Wpf.RenderMode.Skia;
        }

```

**Comment by pauldendulk** [comment]
I did some tests.

In my test setup I added, removed and re-added a layer and did not get the exception you got. 

I did also some tests with the Mapsui code linked to your project. What seems to be happening is that the old MapControl is not disposed. This old MapControl still hold the rendered features in it's canvas, this causes the exception (WPF does not allow a UIElement to be the child of two separate controls).

**Comment by pauldendulk** [comment]
It turned out the WPF MapControl did not properly dispose because it had an event listening to CompositionTarget.Rendering. I solved this by using weak event listeners. 

I can not determine if this fixes the problem in your app because I don't see a dispose in the MapControl.xaml.cs at all, not even if I remove all references to Mapsui. I checked this by adding a finalizer in the MapControl (~MapControl() {}) and write something to Debug.WriteLine there.

**Comment by pauldendulk** [comment]
MapControl is now using weak reference allowing it to be disposed. https://github.com/pauldendulk/Mapsui/releases/tag/1.2.0

**Comment by tkopton** [comment]
Hey guys, looks like you know how to refresh the map using a dedicated method, like the one mentioned:

private void RefreshMap()

This does not work in my code. I can clear the layers but inserting them back in the foreach loop throws an uninitialised exception, even though the array or list has members.
Additionally I cannot find that Method:

this.MainMapControl.Refresh();

There is no Refresh() in MapsUI.map

Sorry for my ignorance, but I am very new to that topic and any help is very much appreciated.

What I would like to achieve is to refresh the map via a method called from another class instance.
Basically there is a UDP server listening for input and any time that server is receiving valid data for the map view I would like the map to be refreshed. The data itself is of course being kept in a model but MVVM doesn't work as here for unknown reason, but this is another story.

**Comment by charlenni** [comment]
Do you call RefreshMap() from the main thread?

**Comment by tkopton** [comment]
No, I would like to call RefreshMap() from a custom service class, that class has a reference to the contentpage instance which is displaying the map. So basically I could call the RefreshMap() method at any time but I have no idea what to do in that method to refresh the content of the map. I could disable and enable layers, this works. But how to refresh the whole content at once?

Another question, maybe you have an idea. 
Using:

mapControl.NativeMap.Layers[i].Enabled = true / false;

I can disable and enable layers individually, that's fine.
But how to disable the map itself to just see the layers with the map background?
The map itself is not a layer, afaik.
My map is a local OSM tile:

public partial class MapView : ContentPage
    {
        private MapsUIView mapControl = new MapsUIView();

        public MapView()
        {
            InitializeComponent();
            mapControl.NativeMap.Layers.Add(OpenStreetMap.CreateTileLayer());
            var layer = GenerateIconLayer();
            mapControl.NativeMap.Layers.Add(layer);
